rpcs3/Emu/Cell/lv2/sys_hid.cpp

```diff
@@ -53,5 +152,36 @@ error_code sys_hid_manager_read(u32 handle, u32 pkg_id, vm::ptr<void> buf, u64 b
 {
 	sys_hid.todo("sys_hid_manager_read(handle=0x%x, pkg_id=0x%x, buf=*0x%x, buf_size=0x%llx)", handle, pkg_id, buf, buf_size);
 
+	if (pkg_id == 2 || pkg_id == 0x81)
+	{
+		sys_hid.trace("sys_hid_manager_read(handle=0x%x, pkg_id=0x%x, buf=*0x%x, buf_size=0x%llx)", handle, pkg_id, buf, buf_size);
+	}
+	else
+	{
+		sys_hid.todo("sys_hid_manager_read(handle=0x%x, pkg_id=0x%x, buf=*0x%x, buf_size=0x%llx)", handle, pkg_id, buf, buf_size);
+	}
+
+	if (pkg_id == 2) {
+		// cellPadGetData
+		// it returns just button array from 'CellPadData'
+		//auto data = vm::static_ptr_cast<u16[64]>(buf);
+		// todo: use handle and dont call cellpad here
+		vm::var<CellPadData> tmpData;
+		if ((cellPadGetData(0, +tmpData) == CELL_OK) && tmpData->len > 0)
+		{
+			u64 cpySize = std::min(static_cast<u64>(tmpData->len) * 2, buf_size * 2);
```

Comment:
why * 2 ?

Reply:
No idea honestly, this was done by other people years ago.

Reply:
It's sizeof(u16), which is the element type if the array.

Reply:
I'd use std::copy instead of memcpy because of this as well.

Reply:
then it should be replaced. also below in the other section

Reply:
Can I get an example here?

Reply:
u64 cpySize = std::min(static_cast<u64>(tmpData->len) * sizeof(u16), buf_size * sizeof(u16));

Reply:
same below